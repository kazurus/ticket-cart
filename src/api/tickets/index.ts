import { mapDestinationsToTickets } from './mappers';
import { destinationsListMock, milePriceMock } from './mockData';
import type { Ticket } from '../../store/tickets/types';

export const getTickets: () => Promise<Ticket[]> = async () =>
  mapDestinationsToTickets(destinationsListMock, milePriceMock);

const tickets = {
  getTickets,
};

export default tickets;
