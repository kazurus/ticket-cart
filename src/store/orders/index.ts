export { ordersActions } from './actions';
export { ordersReducer } from './reducers';
export type { OrdersState } from './reducers';
