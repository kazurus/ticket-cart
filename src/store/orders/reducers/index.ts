import { OrdersActionTypes } from '../actionTypes';
import type { OrdersActions } from '../actionTypes';
import type { OrderId, Order } from '../types';

export type OrdersMapById = Record<OrderId, Order>;

export type OrdersState = {
  entities: {
    byId: OrdersMapById;
    allIds: OrderId[];
  };
};

export const initialState: OrdersState = {
  entities: {
    byId: {},
    allIds: [],
  },
};

// Simple id generation
let nextOrderId = 1;

export function ordersReducer(
  state = initialState,
  action: OrdersActions
): OrdersState {
  switch (action.type) {
    case OrdersActionTypes.ADD_ORDER: {
      const { ticketId } = action.payload;
      const order: Order = {
        // eslint-disable-next-line no-plusplus
        id: nextOrderId++,
        ticketId,
        createdAt: Date.now(),
      };
      const byId = { ...state.entities.byId, [order.id]: order };
      const allIds = state.entities.allIds.concat(order.id);

      return {
        ...state,
        entities: {
          byId,
          allIds,
        },
      };
    }

    case OrdersActionTypes.DELETE_ORDER: {
      const { orderId } = action.payload;
      const { [orderId]: value, ...byId } = state.entities.byId;
      const allIds = state.entities.allIds.filter((id) => id !== orderId);

      return {
        ...state,
        entities: {
          byId,
          allIds,
        },
      };
    }

    default:
      return state;
  }
}
