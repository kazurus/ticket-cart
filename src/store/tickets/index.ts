export { ticketsActions } from './actions';
export { watchLoadTickets } from './sagas';
export { ticketsReducer } from './reducers';
export type { TicketsState } from './reducers';
