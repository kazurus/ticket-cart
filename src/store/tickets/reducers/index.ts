import { TicketsActionTypes, TicketsLoadError } from '../actionTypes';
import type { TicketsAction } from '../actionTypes';
import { TicketId, Ticket } from '../types';

export enum TicketsLoadStatus {
  IDLE = 'IDLE',
  PENDING = 'PENDING',
  READY = 'READY',
  ERROR = 'ERROR',
}

export type TicketsMapById = Record<TicketId, Ticket>;

export type TicketsState = {
  ticketsLoadStatus: TicketsLoadStatus;
  error: TicketsLoadError;
  entities: {
    byId: TicketsMapById;
    allIds: TicketId[];
  };
};

export const initialState: TicketsState = {
  ticketsLoadStatus: TicketsLoadStatus.IDLE,
  error: '',
  entities: {
    byId: {},
    allIds: [],
  },
};

function loadTicketsReducer(
  state: TicketsState,
  action: TicketsAction
): TicketsState {
  switch (action.type) {
    case TicketsActionTypes.FETCH_TICKETS: {
      return {
        ...state,
        ticketsLoadStatus: TicketsLoadStatus.PENDING,
        error: '',
      };
    }

    default:
      return state;
  }
}

function loadTicketsPendingReducer(
  state: TicketsState,
  action: TicketsAction
): TicketsState {
  switch (action.type) {
    case TicketsActionTypes.FETCH_TICKETS_SUCCESS: {
      const { tickets } = action.payload;
      const ticketsMapById = tickets.reduce(
        (ticketsMap, ticket) => ({
          ...ticketsMap,
          [ticket.id]: ticket,
        }),
        {}
      );
      const ticketsIds = tickets.map((ticket) => ticket.id);

      return {
        ...state,
        ticketsLoadStatus: TicketsLoadStatus.READY,
        entities: {
          byId: ticketsMapById,
          allIds: ticketsIds,
        },
      };
    }

    case TicketsActionTypes.FETCH_TICKETS_FAILURE: {
      return {
        ...state,
        ticketsLoadStatus: TicketsLoadStatus.ERROR,
        error: action.payload.error,
      };
    }

    default:
      return state;
  }
}

export function ticketsReducer(
  state: TicketsState = initialState,
  action: TicketsAction
): TicketsState {
  switch (state.ticketsLoadStatus) {
    case TicketsLoadStatus.IDLE:
    case TicketsLoadStatus.ERROR:
      return loadTicketsReducer(state, action);
    case TicketsLoadStatus.PENDING:
      return loadTicketsPendingReducer(state, action);

    default:
      return state;
  }
}
