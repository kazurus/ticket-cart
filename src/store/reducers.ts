import { combineReducers } from 'redux';
import { ticketsReducer } from './tickets';
import type { TicketsState } from './tickets';
import { ordersReducer } from './orders';
import type { OrdersState } from './orders';

export type RootState = {
  tickets: TicketsState;
  orders: OrdersState;
};

export const rootReducer = combineReducers({
  tickets: ticketsReducer,
  orders: ordersReducer,
});
