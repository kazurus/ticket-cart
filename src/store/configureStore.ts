import createSagaMiddleware from 'redux-saga';
import { createStore, applyMiddleware, Middleware, compose } from 'redux';
import { createLogger } from 'redux-logger';

import { rootReducer } from './reducers';
import { rootSaga } from './sagas';

export function configureAppStore() {
  const sagaMiddleware = createSagaMiddleware();

  const middlewares: Middleware[] = [sagaMiddleware];

  const excludeEnv = ['production', 'test'];

  if (!excludeEnv.includes(process.env.NODE_ENV)) {
    middlewares.push(createLogger());
  }

  const composeEnhancer =
    (process.env.NODE_ENV !== 'production' &&
      ((window as any)
        .__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ as typeof compose)) ||
    compose;

  const enhancer = composeEnhancer(applyMiddleware(...middlewares));

  const store = createStore(rootReducer, enhancer);

  sagaMiddleware.run(rootSaga);

  return store;
}
