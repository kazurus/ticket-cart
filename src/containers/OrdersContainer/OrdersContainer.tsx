import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import OrderItem from '../../components/orders/OrderItem';
import OrdersList from '../../components/orders/OrdersList';
import { getOrdersWithTicketData } from '../../store/selectors';
import type { RootState } from '../../store/reducers';
import type { OrderId } from '../../store/orders/types';
import { ordersActions } from '../../store/orders';

const OrdersContainer: React.FC = () => {
  const dispatch = useDispatch();
  const orders = useSelector((state: RootState) =>
    getOrdersWithTicketData(state)
  );

  const deleteOrder = (id: OrderId) => dispatch(ordersActions.deleteOrder(id));

  return (
    <OrdersList>
      {orders.map((order) => (
        <OrderItem
          key={order.id}
          order={order}
          onDelete={() => deleteOrder(order.id)}
        />
      ))}
    </OrdersList>
  );
};

export default OrdersContainer;
