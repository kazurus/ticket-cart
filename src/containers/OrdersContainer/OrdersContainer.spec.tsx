import React from 'react';
import { Store } from 'redux';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureAppStore } from '../../store/configureStore';
import { ticketsActions } from '../../store/tickets';
import { ordersActions } from '../../store/orders';
import { initialState } from '../../store/orders/reducers';
import OrdersContainer from './OrdersContainer';

const renderOrdersContainer = (store: Store) =>
  render(
    <Provider store={store}>
      <OrdersContainer />
    </Provider>
  );

jest.mock('redux-saga/effects', () => {
  const operators = jest.requireActual('redux-saga/effects');
  const delay = jest.fn(() => (s: number) => s);
  return { ...operators, delay };
});

describe('OrdersContainer component', () => {
  let store: ReturnType<typeof configureAppStore>;
  let component: ReturnType<typeof renderOrdersContainer>;

  beforeEach(() => {
    store = configureAppStore();
    component = renderOrdersContainer(store);
    store.dispatch(ticketsActions.loadTickets());
    expect(store.getState().orders).toEqual(initialState);
  });

  afterEach(() => {
    component.unmount();
  });

  it('should render components', () => {
    const { container } = component;
    expect(container).toMatchSnapshot();
  });

  it('should have empty list on mount', () => {
    const { queryAllByTestId } = component;
    expect(queryAllByTestId('order-item')).toHaveLength(0);
  });

  it("shouldn't fetch repos on mount if username is empty", async () => {
    store.dispatch(ordersActions.addOrder(1));

    const { queryAllByTestId } = component;

    expect(queryAllByTestId('order-item')).toHaveLength(1);
  });
});
