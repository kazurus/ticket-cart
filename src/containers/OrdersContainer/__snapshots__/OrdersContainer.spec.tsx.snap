// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersContainer component should render components 1`] = `
<div>
  <div>
    <h3>
      Orders
    </h3>
    <div />
  </div>
</div>
`;
